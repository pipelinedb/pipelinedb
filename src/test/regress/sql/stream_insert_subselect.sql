CREATE SCHEMA test_stream_insert_subselect;
SET search_path TO test_stream_insert_subselect,public;

CREATE TABLE t (x integer);
INSERT INTO t (SELECT generate_series(1, 100));

CREATE CONTINUOUS VIEW v0 AS SELECT x::integer FROM stream;

INSERT INTO stream (x) (SELECT * FROM t);
INSERT INTO stream (x) (SELECT * FROM (SELECT x AS y FROM t) s0);

SELECT COUNT(DISTINCT x) FROM v0;

CREATE CONTINUOUS VIEW v1 AS SELECT x::integer, COUNT(*) FROM stream GROUP BY x;
INSERT INTO stream (x) (SELECT generate_series(1, 1000));

SELECT COUNT(DISTINCT x) FROM v1;

-- It's not possible to SELECT from another stream in a stream INSERT
CREATE STREAM s0 (x integer);
INSERT INTO stream (x) (SELECT x FROM s0);

CREATE CONTINUOUS VIEW v2 AS SELECT x::integer FROM stream;
INSERT INTO stream (x) (SELECT generate_series(1, 1000) AS x ORDER BY random());

SELECT COUNT(DISTINCT x) FROM v2;

CREATE CONTINUOUS VIEW v3 AS SELECT x::integer FROM stream;
INSERT INTO stream (x) (SELECT * FROM t WHERE x IN (SELECT generate_series(1, 20)));

SELECT * FROM v3 ORDER BY x;

CREATE CONTINUOUS VIEW v4 AS SELECT COUNT(*) FROM stream;
INSERT INTO stream (price, t) SELECT 10 + random() AS price, current_timestamp - interval '1 minute' * random() AS t FROM generate_series(1, 1000);

SELECT * FROM v4;

DROP SCHEMA test_stream_insert_subselect CASCADE;
