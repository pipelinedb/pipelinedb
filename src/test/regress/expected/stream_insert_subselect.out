CREATE TABLE stream_subselect_t (x integer);
INSERT INTO stream_subselect_t (SELECT generate_series(1, 100));
CREATE CONTINUOUS VIEW stream_subselect_v0 AS SELECT x::integer FROM stream_subselect_stream;
INSERT INTO stream_subselect_stream (x) (SELECT * FROM stream_subselect_t);
INSERT INTO stream_subselect_stream (x) (SELECT * FROM (SELECT x AS y FROM stream_subselect_t) s0);
SELECT COUNT(DISTINCT x) FROM stream_subselect_v0;
 count 
-------
   100
(1 row)

CREATE CONTINUOUS VIEW stream_subselect_v1 AS SELECT x::integer, COUNT(*) FROM stream_subselect_stream GROUP BY x;
INSERT INTO stream_subselect_stream (x) (SELECT generate_series(1, 1000));
SELECT COUNT(DISTINCT x) FROM stream_subselect_v1;
 count 
-------
  1000
(1 row)

-- It's not possible to SELECT from another stream in a stream INSERT
CREATE STREAM stream_subselect_s0 (x integer);
INSERT INTO stream_subselect_stream (x) (SELECT x FROM stream_subselect_s0);
ERROR:  "stream_subselect_s0" is a stream
HINT:  Streams can only be read by a continuous view's FROM clause.
CREATE CONTINUOUS VIEW stream_subselect_v2 AS SELECT x::integer FROM stream_subselect_stream;
INSERT INTO stream_subselect_stream (x) (SELECT generate_series(1, 1000) AS x ORDER BY random());
SELECT COUNT(DISTINCT x) FROM stream_subselect_v2;
 count 
-------
  1000
(1 row)

CREATE CONTINUOUS VIEW stream_subselect_v3 AS SELECT x::integer FROM stream_subselect_stream;
INSERT INTO stream_subselect_stream (x) (SELECT * FROM stream_subselect_t WHERE x IN (SELECT generate_series(1, 20)));
SELECT * FROM stream_subselect_v3 ORDER BY x;
 x  
----
  1
  2
  3
  4
  5
  6
  7
  8
  9
 10
 11
 12
 13
 14
 15
 16
 17
 18
 19
 20
(20 rows)

CREATE CONTINUOUS VIEW stream_subselect_v4 AS SELECT COUNT(*) FROM stream_subselect_stream;
INSERT INTO stream_subselect_stream (price, t) SELECT 10 + random() AS price, current_timestamp - interval '1 minute' * random() AS t FROM generate_series(1, 1000);
SELECT * FROM stream_subselect_v4;
 count 
-------
  1000
(1 row)

DROP CONTINUOUS VIEW stream_subselect_v0;
DROP CONTINUOUS VIEW stream_subselect_v1;
DROP CONTINUOUS VIEW stream_subselect_v2;
DROP CONTINUOUS VIEW stream_subselect_v3;
DROP CONTINUOUS VIEW stream_subselect_v4;
DROP TABLE stream_subselect_t;
